fix memory leak in CompilerServiceBase by removing event handler in dispose

campersau opened the pull request:
`ICompilerService` now implements `IDisposable` so all implementing classes can be properly disposed.
This is required to prevent a memory leak within `CompilerServiceBase` which previously did not unregister the `AssemblyResolve` event handler.
Also the `Dispose` method of `DirectCompilerServiceBase` was never called which disposes the CodeDomProvider. (There was also a nice comment: `Are we explicily disposing of this instance?`)

Here is a screenshot from .NET Memory Profile before this fix:
![image](https://cloud.githubusercontent.com/assets/4009570/15647423/c425b6d4-2663-11e6-8d12-b3873136fcb1.png)
The red line are the CompilerServices.
The orange line are the CodeDomProviders from `DirectCompilerServiceBase`.

After this fix these objects were properly gced.

matthid commented:
I don't think it is that easy. The resolve event needs to be handled as long as instances and types from the loaded assemblies are used. This however is forever as we cannot unload a assembly once it has been loaded (maybe in the future with .net core).

matthid commented:
But I agree that there is a problem here...

campersau commented:
I have debugged this quite a bit but could come up with a case where the AssemblyResolve was called after CompileType...
Do you have a test case for that?
Also I thought that the IReferenceResolver was there to do that and not the CompilerService. But maybe I do not understand the code enough...

matthid commented:
Yes the one that is failing right now. I think this can only be reproduced on windows as mono loads assemblies a bit stricter. 

you are correct though, logically i would say it's no the responsibilty of the compiler service. The code is quite unclean in that regard. Thats what I meant with there is a problem here...

campersau commented:
Ah, now I see it.
So one option might be to move the AssemblyResolve resposibility into the IReferenceResolver. Then only one AssemblyResolve handler would be added and we can dispose the CompilerService after CompileType.

matthid commented:
Yeah. Changes here are quite error prone though. On the other hand we have a pretty good test coverage so feel free to make some radical changes if you need to. I try to carefuly review them.

campersau commented:
Maybe something like this?
Works also for IsolatedRazorEngine (no tests for that at the moment).

campersau commented:
Also added test for dynamic assembly loading inside of IsolatedRazorEngineService.

I moved the assembly loading logic to the RazorEngineService for now because of backward compatibility. WDYT?

This was very big problem for us. After two days our servers hit their memory limit of 16 GB because of this. Now our servers are running over almost a week with this fix and memory consumption is moderate.

matthid commented on src/source/RazorEngine.Core/Compilation/ReferenceResolver/CompilerReference.cs:
```diff
@@ -130,6 +130,35 @@ public static CompilerReference From(byte[] byteArray)
+        /// <param name="assemblyName">name of the assembly to resolve</param>
+        /// <param name="references">references to check</param>
+        /// <returns>the resolved assembly or null</returns>
+        public static Assembly Resolve(string assemblyName, IEnumerable<CompilerReference> references)
```
Does this need to be public?

matthid commented on src/source/RazorEngine.Core/Compilation/CompilerServiceBase.cs:
```diff
@@ -438,6 +412,34 @@ protected virtual void Inspect(CodeCompileUnit unit)
+        /// <summary>
+        /// Cleans up the <see cref="CompilerServiceBase"/> instance.
+        /// </summary>
+        ~CompilerServiceBase()
```
I don't think we need a finalizer here, since we don't have any unmanaged stuff ourself.

matthid commented:
Yeah I think the patch is really nice, thanks! Just two smaller things in-line. (Maybe I'm wrong on the finalizer thing, but I think unmanaged resources are already cleaned up by the finalizer of the CodeDom instance if needed.)

matthid commented on src/source/RazorEngine.Core/Templating/RazorEngineService.cs:
```diff
@@ -48,6 +51,12 @@ internal RazorEngineService(ITemplateServiceConfiguration config)
+
+        private Assembly AssemblyResolve(object sender, ResolveEventArgs args)
+        {
+            return CompilerReference.Resolve(args.Name, _config.ReferenceResolver.GetReferences(null));
```
This might be a bit dangerous. I think custom implementation of `IReferenceResolver` might not behave as expected when giving `null`. Ideally we would "save" the references once we retrieved them in the compilation process and them use the saved list for the `AssemblyResolve` event... But I think we could do that in a later PR. If you don't want to include it here please open an issue :).

campersau commented:
Changed according to feedback.

I also moved the assembly resolve logic to RazorEngineCore, because there we have access to the TypeContext which we can (mis)use to get the CompilerReferences. Which I think is a bit better instead of calling `GetReferences(null)`.

matthid commented on src/source/RazorEngine.Core/Compilation/TypeContext.cs:
```diff
@@ -56,6 +57,11 @@ internal TypeContext(string className, ISet<string> namespaces)
+        /// <summary>
+        /// Gets or sets the compiler references
+        /// </summary>
+        public ISet<CompilerReference> References { get; set; }
```
On the one hand I think this should be internal. On the other hand the `ICompilerService` is an extension point so people should be able to do the same than the 'out of the box' implementation (in fact there is `RazorEngine.Roslyn`...). So there needs to be a point to add you own references. I think public is fine for now. (As long as it is undocumented it's kind of 'use it on your own risk').

matthid commented:
Regarding the 'misuse'. Yeah the whole API is kind of a mess and I have no idea how to navigate out of it with small improvements. It always seems to backflip and keeps getting worse. On the first hand everything seems to be logically independent, but actually everything is quite coupled though non obvious and 'unspoken' dependencies...

Thanks for taking care of this!
